### ISSUE TITLE
Add BlocProxyProvider feature

### ISSUE THREAD

**Issue by jesusrp98** [felangel/bloc#2124]
**Is your feature request related to a problem? Please describe.**
At this time, you can't really make blocs dependent on other blocs. The implementation of this feature, already implemented on the `provider` package, would be very beneficial for all developers & use cases.

**Describe the solution you'd like**
Create a new 'BlocProvider' called `BlocProxyProvider` that would work exactly the same as a regular`ProxyProvider`.

The documentation to the `ProxyProvider` could be found [here](https://pub.dev/packages/provider#proxyprovider).

**Comment by felangel** [comment]
Hi @jesusrp98 👋
Thanks for opening an issue!

Can you please provide a concrete use case? Thanks!

**Comment by jesusrp98** [comment]
Here's a simple example where you could use this feature:

```
MultiProvider(
      providers: [
        ChangeNotifierProvider(
          create: (_) => StoresCubit(StoresService()),
        ),
        ChangeNotifierProxyProvider(
          create: (_) => BooksCubit(),
          update: (_, StoresCubit storesCubit, __) =>
              BooksCubit(
            BooksService(),
            stores: storesCubit.value,
          ),
        ),
      ],
    ),
```

In this case we have a cubit dedicated for handling a series of book stores, data that we'd need inside our `BooksCubit`. This allows a 'dependency injection` from the `StoresCubit` quite easy.

As I said before, this is already available inside the `provider` package, and it would be very beneficial to expose this functionality from that package into `flutter_bloc`.

**Comment by jjoelson** [comment]
@jesusrp98 When the `BooksCubit` is updated, do your `BlocBuilders` further down the hierarchy start using the new instance? I just posted [this issue](https://github.com/felangel/bloc/issues/2127) because I'm trying to do something similar and I'm seeing `BlocBuilder` continue to use the old instance.

**Comment by jesusrp98** [comment]
Yes @jjoelson , this is how it works with `provider`. This update of `BooksCubit` will be propagated through out the whole widget tree.

**Comment by jjoelson** [comment]
I know that's how `provider` works, but [`_BlocBuilderBaseState`](https://github.com/felangel/bloc/blob/daf09b994cbcc4b7105572d954c069e4795abbd0/packages/flutter_bloc/lib/src/bloc_builder.dart#L124) (which is what `BlocBuilder` uses) retrieves the cubit from the context using `context.read<C>()`:
```
  @override
  void initState() {
    super.initState();
    _cubit = widget.cubit ?? context.read<C>();
    _state = _cubit.state;
  }

  @override
  void didUpdateWidget(BlocBuilderBase<C, S> oldWidget) {
    super.didUpdateWidget(oldWidget);
    final oldCubit = oldWidget.cubit ?? context.read<C>();
    final currentCubit = widget.cubit ?? oldCubit;
    if (oldCubit != currentCubit) {
      _cubit = currentCubit;
      _state = _cubit.state;
    }
  }
```

This means `BlocBuilder` doesn't get updated when proxy provider updates the cubit. [If I instead change it to use `context.watch<C>()` in the build method](https://github.com/jjoelson/bloc/commit/00ebc9eb0e92bf354d509f6ef34351cf5c95e382), then it works:
```
  @override
  Widget build(BuildContext context) {
    _cubit = widget.cubit ?? context.watch<C>();
    _state = _cubit.state;

    return BlocListener<C, S>(
      cubit: _cubit,
      listenWhen: widget.buildWhen,
      listener: (context, state) => setState(() => _state = state),
      child: widget.build(context, _state),
    );
  }
```

I filed this as a bug yesterday, but I'm not actually sure if it's a bug or intended behavior.

**Comment by felangel** [comment]
@jjoelson @jesusrp98 I've opened https://github.com/felangel/bloc/pull/2181 to fix the issue of `BlocBuilder` not reacting to changes in the provided bloc instance. With this change it should be possible to use `ProxyProvider` from `package:provider` to achieve the desired behavior. I would prefer not to create a `BlocProxyProvider` because I don't believe there is anything bloc-specific in this case and developers can pull in `package:provider` as needed. Let me know what you think and sorry for the delayed response! 🙏

**Comment by jesusrp98** [comment]
It's great that this fix solves the issue regarding `BlocBuilder` not updating itself. And if this also makes compatible a `bloc` instance with a `ProxyProvider` with the same behaviour as a regular provider I'm also happy!

**Comment by felangel** [comment]
Closing this since it will be addressed by #2127
